### ISSUE TITLE
testscript: add a "don't care about success" operator

### ISSUE THREAD

**Issue by myitcv** [rogpeppe/go-internal#93]
Per https://go-review.googlesource.com/c/go/+/223746

However per @rogpeppe on Slack, this would be a backwards incompatible change because the signature of [`Params.Cmds`](https://pkg.go.dev/github.com/rogpeppe/go-internal/testscript?tab=doc#Params) would need to change.

Options:

* Add a new field
* ...

cc @rogpeppe @mvdan

**Comment by mvdan** [comment]
I have a somewhat pressing need for this - Go 1.15 (master) is making the default `-h` flag report an exit status of 0 instead of the old 1, so now all of my `! mytool -h` lines are failing. I could use conditions and split the line into multiple, but a simpler fix would be to use `? mytool -h` instead, as I only care that the flag prints what I expect it to print.

I agree that a new field, deprecating the current `Cmds` (but keeping it working), would probably be the safest bet. This is similar to what Robert did with https://golang.org/pkg/go/types/#Importer - see ImporterFrom.

Something he did there, which we could emulate, is to add an extra `mode` integer parameter, so that future knobs may be added without yet another deprecation. I guess we could do that here, and replace both `neg bool` and the new `want simpleStatus` with two bits in a new enum-like mode. Then, the signature could end up like:

```
type CmdMode uint32

const (
	CmdWantFailure CmdMode = 1 << iota
	CmdWantEither
)
[...]
	Cmds map[string]func(ts *TestScript, mode CmdMode, args []string)
```

**Comment by mvdan** [comment]
I should add - `CmdWantFailure | CmdWantEither` should never happen, since they are exclusive knobs.

**Comment by myitcv** [comment]
> Go 1.15 (master) is making the default `-h` flag report an exit status of 0 instead of the old 1

I'm assuming the implication of this change for the likes of `testscript` and other testers has been discussed? This feels like a fairly significant change...

Agree with your suggestion re modes BTW.

**Comment by twpayne** [comment]
https://go-review.googlesource.com/c/go/+/223746 is now merged. Maybe this issue can be closed.

**Comment by myitcv** [comment]
@twpayne it's precisely because it's been merged that we need this issue 😄  - i.e. we need to "keep up to date" with the upstream.

**Comment by twpayne** [comment]
I had a look at doing this. There's some non-trivial divergence between upstream and this package now. The easiest way to resolve this is to probably copy in from upstream again. The changes are then fairly small, mainly using exported methods instead of private methods. Otherwise, you have to pick through quite a lot of code. However, this will lead to the loss of any changes made to this package.

**Comment by twpayne** [comment]
I need this again.

As an alternative to the `?` operator (which, I agree, is the correct way to do it), how about adding a new verb `execdontcareaboutsuccess` verb, which, er, does exactly what it says on the tin, so you can write:

```
execdontcareaboutsuccess mytool -h
```

This has the advantage of being backwards-compatible, straightforward to implement, and providing the desired functionality.

**Comment by thepudds** [comment]
Hi @twpayne, could you implement this yourself as small [custom command](https://pkg.go.dev/github.com/rogpeppe/go-internal/testscript#RunMain) that is specific to your project?

That might then serve as data for what to do about this issue.

**Comment by twpayne** [comment]
> Hi @twpayne, could you implement this yourself as small [custom command](https://pkg.go.dev/github.com/rogpeppe/go-internal/testscript#RunMain) that is specific to your project?

I don't think so because the [existing implementation of `exec`](https://github.com/rogpeppe/go-internal/blob/ec119421bb970ac974ea4ac16d8b58ffd0d727fc/testscript/cmd.go#L226-L270) uses unexported functions like `ts.findBackground` and `ts.exec`. I already have a [lot of custom commands](https://github.com/twpayne/chezmoi/blob/7edfed76b54dcda466cd803a48ca311c818f9009/pkg/cmd/main_test.go#L53-L74).

**Comment by mvdan** [comment]
FWIW in CUE we did a variation of what @twpayne suggested - just for one command, but it's enough to get us unblocked:

```
func TestMain(m *testing.M) {
	os.Exit(testscript.RunMain(m, map[string]func() int{
		"cue": MainTest,
		// Until https://github.com/rogpeppe/go-internal/issues/93 is fixed,
		// or we have some other way to use "exec" without caring about success,
		// this is an easy way for us to mimic `? exec cue`.
		"cue_exitzero": func() int {
			MainTest()
			return 0
		},
[...]
```
